### PR TITLE
MAINT:update boost to fix `skewnorm.ppf`

### LINKED ISSUE #20124: BUG: stats.skewnorm.ppf returns wrong values with moderately high skew parameters
### Describe your issue.

I'm trying to compute the inverse CDF (i.e. PPF) of several skewnorm distributions. I am getting nonsense answers when the skew parameter is moderately large.

### Reproducing Code Example

```python
from scipy.stats import halfnorm, skewnorm


#  This is the correct ppf for a standard skewnorm with skew 500
q = 0.012533469508013

#  As confirmation of the above result, the skewnorm with skew 500 is
#  extremely close to a halfnorm distribution.
#  In fact, the ppf of the halfnorm agrees with q to 13 decimal places.
halfnorm.ppf(0.01, 0, 1)
#  0.012533469508069276

#  This should theoretically be 0.01, and it is very close.
skewnorm.cdf(q, 500, 0, 1)
#  0.009999999999999898

#  This should theoretically be q, but it's not at all.
#  This is the problem.
skewnorm.ppf(0.01, 500, 0, 1)
#  4.833181656374989e+142
```

### Error message

```shell
/opt/conda/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:9141: RuntimeWarning: overflow encountered in _skewnorm_ppf
  return _boost._skewnorm_ppf(x, 0, 1, a)
```

### SciPy/NumPy/Python version and system information

```shell
1.12.0 1.26.3 sys.version_info(major=3, minor=11, micro=7, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/conda/include
    lib directory: /opt/conda/lib
    name: blas
    openblas configuration: unknown
    pc file directory: /opt/conda/lib/pkgconfig
    version: 3.9.0
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/conda/include
    lib directory: /opt/conda/lib
    name: lapack
    openblas configuration: unknown
    pc file directory: /opt/conda/lib/pkgconfig
    version: 3.9.0
  pybind11:
    detection method: pkgconfig
    include directory: /opt/conda/include
    name: pybind11
    version: 2.11.1
Compilers:
  c:
    args: -march=nocona, -mtune=haswell, -ftree-vectorize, -fPIC, -fstack-protector-strong,
      -fno-plt, -O2, -ffunction-sections, -pipe, -isystem, /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix, -march=nocona, -mtune=haswell,
      -ftree-vectorize, -fPIC, -fstack-protector-strong, -fno-plt, -O2, -ffunction-sections,
      -pipe, -isystem, /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix, -DNDEBUG, -D_FORTIFY_SOURCE=2,
      -O2, -isystem, /opt/conda/include, -DNDEBUG, -D_FORTIFY_SOURCE=2, -O2, -isystem,
      /opt/conda/include
    commands: /home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/_build_env/bin/x86_64-conda-linux-gnu-cc
    linker: ld.bfd
    linker args: -Wl,-O2, -Wl,--sort-common, -Wl,--as-needed, -Wl,-z,relro, -Wl,-z,now,
      -Wl,--disable-new-dtags, -Wl,--gc-sections, -Wl,--allow-shlib-undefined, -Wl,-rpath,/opt/conda/lib,
      -Wl,-rpath-link,/opt/conda/lib, -L/opt/conda/lib, -Wl,-O2, -Wl,--sort-common,
      -Wl,--as-needed, -Wl,-z,relro, -Wl,-z,now, -Wl,--disable-new-dtags, -Wl,--gc-sections,
      -Wl,--allow-shlib-undefined, -Wl,-rpath,/opt/conda/lib, -Wl,-rpath-link,/opt/conda/lib,
      -L/opt/conda/lib, -march=nocona, -mtune=haswell, -ftree-vectorize, -fPIC, -fstack-protector-strong,
      -fno-plt, -O2, -ffunction-sections, -pipe, -isystem, /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix, -march=nocona, -mtune=haswell,
      -ftree-vectorize, -fPIC, -fstack-protector-strong, -fno-plt, -O2, -ffunction-sections,
      -pipe, -isystem, /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix, -DNDEBUG, -D_FORTIFY_SOURCE=2,
      -O2, -isystem, /opt/conda/include, -DNDEBUG, -D_FORTIFY_SOURCE=2, -O2, -isystem,
      /opt/conda/include
    name: gcc
    version: 12.3.0
  c++:
    args: -fvisibility-inlines-hidden, -fmessage-length=0, -march=nocona, -mtune=haswell,
      -ftree-vectorize, -fPIC, -fstack-protector-strong, -fno-plt, -O2, -ffunction-sections,
      -pipe, -isystem, /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix, -fvisibility-inlines-hidden,
      -fmessage-length=0, -march=nocona, -mtune=haswell, -ftree-vectorize, -fPIC,
      -fstack-protector-strong, -fno-plt, -O2, -ffunction-sections, -pipe, -isystem,
      /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix, -DNDEBUG, -D_FORTIFY_SOURCE=2,
      -O2, -isystem, /opt/conda/include, -DNDEBUG, -D_FORTIFY_SOURCE=2, -O2, -isystem,
      /opt/conda/include
    commands: /home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/_build_env/bin/x86_64-conda-linux-gnu-c++
    linker: ld.bfd
    linker args: -Wl,-O2, -Wl,--sort-common, -Wl,--as-needed, -Wl,-z,relro, -Wl,-z,now,
      -Wl,--disable-new-dtags, -Wl,--gc-sections, -Wl,--allow-shlib-undefined, -Wl,-rpath,/opt/conda/lib,
      -Wl,-rpath-link,/opt/conda/lib, -L/opt/conda/lib, -Wl,-O2, -Wl,--sort-common,
      -Wl,--as-needed, -Wl,-z,relro, -Wl,-z,now, -Wl,--disable-new-dtags, -Wl,--gc-sections,
      -Wl,--allow-shlib-undefined, -Wl,-rpath,/opt/conda/lib, -Wl,-rpath-link,/opt/conda/lib,
      -L/opt/conda/lib, -fvisibility-inlines-hidden, -fmessage-length=0, -march=nocona,
      -mtune=haswell, -ftree-vectorize, -fPIC, -fstack-protector-strong, -fno-plt,
      -O2, -ffunction-sections, -pipe, -isystem, /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix, -fvisibility-inlines-hidden,
      -fmessage-length=0, -march=nocona, -mtune=haswell, -ftree-vectorize, -fPIC,
      -fstack-protector-strong, -fno-plt, -O2, -ffunction-sections, -pipe, -isystem,
      /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix, -DNDEBUG, -D_FORTIFY_SOURCE=2,
      -O2, -isystem, /opt/conda/include, -DNDEBUG, -D_FORTIFY_SOURCE=2, -O2, -isystem,
      /opt/conda/include
    name: gcc
    version: 12.3.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.7
  fortran:
    args: -march=nocona, -mtune=haswell, -ftree-vectorize, -fPIC, -fstack-protector-strong,
      -fno-plt, -O2, -ffunction-sections, -pipe, -isystem, /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix
    commands: /home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/_build_env/bin/x86_64-conda-linux-gnu-gfortran
    linker: ld.bfd
    linker args: -Wl,-O2, -Wl,--sort-common, -Wl,--as-needed, -Wl,-z,relro, -Wl,-z,now,
      -Wl,--disable-new-dtags, -Wl,--gc-sections, -Wl,--allow-shlib-undefined, -Wl,-rpath,/opt/conda/lib,
      -Wl,-rpath-link,/opt/conda/lib, -L/opt/conda/lib, -Wl,-O2, -Wl,--sort-common,
      -Wl,--as-needed, -Wl,-z,relro, -Wl,-z,now, -Wl,--disable-new-dtags, -Wl,--gc-sections,
      -Wl,--allow-shlib-undefined, -Wl,-rpath,/opt/conda/lib, -Wl,-rpath-link,/opt/conda/lib,
      -L/opt/conda/lib, -march=nocona, -mtune=haswell, -ftree-vectorize, -fPIC, -fstack-protector-strong,
      -fno-plt, -O2, -ffunction-sections, -pipe, -isystem, /opt/conda/include, -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/scipy-split_1705882420314/work=/usr/local/src/conda/scipy-split-1.12.0,
      -fdebug-prefix-map=/opt/conda=/usr/local/src/conda-prefix
    name: gcc
    version: 12.3.0
  pythran:
    include directory: ../../_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/conda/bin/python
  version: '3.11'
```

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Supersedes #20125 
Closes #20124

#### What does this implement/fix?
The skewnorm quantile calculation was fixed upstream, so this updates boost to the latest commit of the development branch.

#### Additional information
CC @maresb in case you have more stress tests for the skewnorm distribution.

I added the backport-candidate label as this is a bug fix but feel free to ignore @tylerjereddy as boost had many moving parts lately. Might be worth to run wheel tests as boost was prone to subtle bugs, will leave the decision to people more involved in that though.

### CODE REVIEW

**Comment by maresb** [comment]
Thanks a lot @dschmitz89 for following up on this! I commented in https://github.com/boostorg/math/issues/1120#issuecomment-2094793170, but I'm hopeful that your work here will be helpful for my current applications.

The way scipy patches Boost's skewnorm CDF [here](https://github.com/scipy/scipy/blob/4a537b736203190a5bca1d00f7e2659a507e1ad3/scipy/stats/_continuous_distns.py#L9278) is really helpful. But unfortunately if we want to use Boost's PPF solver, that will use Boost's unstable CDF function, so it will be plagued by the upstream inaccuracies.

```python
from scipy.stats import skewnorm
from scipy.stats._boost import _skewnorm_cdf
from scipy import __version__

__version__  # '1.10.0'
skewnorm.cdf(-3, 5)  # 4.136670779725204e-55
_skewnorm_cdf(-3, 0, 1, 5)  # 2.3852447794681098e-18
```

With Mathematica, `N[CDF[SkewNormalDistribution[0, 1, 5], -3], 20]` gives `4.1366707600635155404e-55`, so we see that scipy's numerical integration is very close, while Boost is way off.

I think I know how to stably compute the CDF without numerical integration, but I'm not sure. I'd need to first implement it and test it thoroughly. I'd like to contribute it upstream to Boost and get rid of the numerical integration fallback here, but I'm not sure when I'll be able to get around to it.

**Comment by dschmitz89** [comment]
Fixing boost's CDF would be very helpful, thanks for looking into that still. Do you agree that the update for the PPF here is an improvement already?

**Comment by maresb** [comment]
I don't have the time to investigate in-depth, but you seem to have fixed my original test case to fairly good precision. Barring any regressions, which hopefully the test suite mostly prevents, this appears to me like a substantial improvement! :tada: 

Thanks again @dschmitz89!

**Comment by maresb** [comment]
For my current application I only need to determine the PPF of roughly the first percentile. Boost's CDF should be safe as long as the quantile can be expressed as the difference between two floats of order `1`. I think that's what this achieves. It'd be really excellent if this works! :crossed_fingers:

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:4013]:
```suggestion
        # gh-20124 reported that Boost's ppf was wrong for high skewness
```

**Comment by dschmitz89** [comment]
I went ahead and tested the wheels: all pass, one test failure is unrelated.

**Comment by mdhaber** [comment]
Is it the change to Boost that made the wheel builder tests happen?

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:4015]:
```suggestion
        # InverseCDF[SkewNormalDistribution[0, 1, 500], 0.01] in Wolfram Alpha.
```

Comment (maresb) [reply scipy/stats/tests/test_distributions.py:4015]:
```suggestion
        # N[InverseCDF[SkewNormalDistribution[0, 1, 500], 1/100], 14] in Wolfram Alpha.
```

**Comment by mdhaber** [comment]
@dschmitz89 This update seems to fix gh-19348 on my end, too. Could you check on your side, and if so add a test?

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:4014]:
```suggestion
        # Reference value was calculated using
```

**Comment by dschmitz89** [comment]
> @dschmitz89 This update seems to fix [gh-19348](https://github.com/scipy/scipy/issues/19348) and [gh-18906](https://github.com/scipy/scipy/issues/18906) on my end, too (or they were already fixed). Could you check on your side, and if so add a test?
> 
> _Update - seems like they were already fixed by something? I'm closing those issues. If you can confirm locally that they appear to be resolved by running the tests mentioned in my last posts (in each issue), I'd appreciate it!_

Good point, I had already forgotten aboutthose. I can confirm that both issues are resolved in main, probably due to the Boost update in https://github.com/scipy/scipy/pull/20393 .

**Comment by mdhaber** [comment]
My only hesitation with merging here is the submodule update, since people tend to find it disruptive and it was already updated quite recently. Would you like this to merge now, or can it wait until release time (at which point we might update to latest Boost)?

**Review by mdhaber — APPROVED** [review]

**Comment by dschmitz89** [comment]
I agree. For the sake of not annoying contributors shortly before a release, let's delay until after the 1.14 release.

**Comment by mdhaber** [comment]
Oh, actually, I still wanted this to make it into 1.14. Let's just ask others who might have an opinion:

@tylerjereddy @rgommers this update to the Boost submodule would close gh-20124. There was just an update to the Boost submodule in gh-20393, and I know we don't like submodule updates to be too frequent. When would be the best time to merge this - now, wait a bit but merge before 1.14 branches, or wait until after 1.14 branches?

**Comment by rgommers** [comment]
"it fixes a bug" is a good enough reason to bump a submodule I'd say.

**Comment by tylerjereddy** [comment]
The backport label has been removed because on MacOS ARM with clang-16.0.6 I saw additional test failures with this branch cherry-picked in to the release branch (not just for the new test added here).

**Comment by maresb** [comment]
Hey, great news, I just wanted to mention that I've been using a dev version of scipy from right after this merge. Previously I experienced a failure rate of roughly one in a million from my dataset. Now, after billions of invocations, I haven't seen a single failure. This is awesome, thanks so much everyone!!!

**Pull request merged**
